### PR TITLE
#556: update PATH on pc run

### LINKED ISSUE #556: ProcessContext should compute PATH on run and not in constructor
As a IDEasy developer, I want to use `ProcessContext` seamlessly so that I do not cause bugs if I do not know its implementation details.

The problem started in PR #515 with this code:
```java
      if (existsEnvironmentContext) {
        pc = this.context.newProcess().errorHandling(ProcessErrorHandling.THROW).executable(binaryPath).addArgs(args);
        install(pc, true);
      } else {
        install(true);
        pc = this.context.newProcess().errorHandling(ProcessErrorHandling.THROW).executable(binaryPath).addArgs(args);
      }
```

I left a review comment that we should move the duplicated code to create the `ProcessContext` before the `if` statement to avoid redundancy.
However, what we then figured out is the following:

`this.context.newProcess()` is creating a new `ProcessContextImpl` and its constructor is computing all environment variables and setting them in the `ProcessBuilder`. Doing this at construction time in general makes sense since we might use `withEnvVar` method to override some environment variable.

However, the when the environment variables are evaluated and set, this also includes the `PATH` variable:
https://github.com/devonfw/IDEasy/blob/c6c7d23b852454eda440a51a393cdd36f849d354/cli/src/main/java/com/devonfw/tools/ide/variable/IdeVariables.java#L24
Therefore the recent path is evaluated from the `SystemPath`:
https://github.com/devonfw/IDEasy/blob/c6c7d23b852454eda440a51a393cdd36f849d354/cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java#L231

However, that value (can) change[s] during the installation of tools because if a tool gets newly installed a new folder like `software/«tool»/bin` may be created and therefore added to `SystemPath` and therefore end up in the value of `PATH`.
So if `install` is called after the `ProcessContext` is created, we can cause a bug.

Even worse with the tool-dependencies approach we now have to create the `ProcessContext` before in order to pass it to the `install` method and "tweak" it in there (e.g. for `tomcat` the variable `CATALINA_HOME` may be set and because `tomcat` depends on `java` it will delegate installation of a compatible version of `java` and therefore set `JAVA_HOME` to that installation).

As a result we IMHO need to handle the `PATH` variable in a special way and when we call `ProcessContext.run()` we should compute the latest `PATH` and set it as environment variable in the `ProcessBuilder`. IMHO it is simpler to just update the `PATH` in `run()` and therefore "compute" it two times: at construction time and then again in `run()`. This does not harm since we manage the `PATH` efficiently via `SystemPath` and actively update it from `install` so we do not have to do directory traversals to compute the `PATH` every time (only happens when IDEasy is launched and context is created).

Theoretically some case could happen, where someone creates a `ProcessContext` and wants to set a custom `PATH` via `withEnvVar` and that would then stop working because we overwrite this later in the `run` method. If somebody wants to solve this story with perfection, he could track this case and then prevent the auto-update, but it is also fine if we ignore this until this problem may ever happen and then create another bug issue for it.

### PR DESCRIPTION
**Pull request by hohwille**
fixes #556 
* add `withPathEntry(Path)` method for future implementations of `setEnvironment` methods
* add ability to create a modified `SystemPath` via `withPath` method
* change ProcessContext to track PATH changes and apply them only in `run` method
* also move the code to find the binary executable to `run` method using the final PATH
* improved logging
* changed `ToolCommandlet` to simplify after this "bug" has been fixed

### CODE REVIEW

**Review by jan-vcapgemini — CHANGES_REQUESTED** [review]
LGTM. Added some small CRs.

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java [hunk 1]
```diff
@@ -246,6 +258,15 @@ public String toString(WindowsPathSyntax pathSyntax) {
     return sb.toString();
   }
 
+  public SystemPath withPath(String overriddenPath, List<Path> extraPathEntries) {
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:269]: Done.

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java [hunk 2]
```diff
@@ -86,14 +89,9 @@ public SystemPath(IdeContext context, String envPath, Path ideRoot, Path softwar
    * @param softwarePath the {@link IdeContext#getSoftwarePath() software path}.
    * @param pathSeparator the path separator char (';' for Windows and ':' otherwise).
    */
-  public SystemPath(IdeContext context, String envPath, Path ideRoot, Path softwarePath, char pathSeparator) {
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:89]: Done.

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java [hunk 3]
```diff
@@ -74,7 +77,7 @@ public SystemPath(IdeContext context, String envPath) {
    */
   public SystemPath(IdeContext context, String envPath, Path ideRoot, Path softwarePath) {
 
-    this(context, envPath, ideRoot, softwarePath, File.pathSeparatorChar);
+    this(context, envPath, ideRoot, softwarePath, File.pathSeparatorChar, Collections.emptyList());
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:80]: What is wrong here? IMHO false positive.

[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:80]: Jup, was a false positive, sorry.

**Review by jan-vcapgemini — APPROVED** [review]
Ready to merge.

**Pull request merged**